crates/solver/src/liquidity.rs

```diff
@@ -352,19 +352,16 @@ impl std::fmt::Debug for StablePoolOrder {
     }
 }
 
-pub fn token_pairs<T>(reserves: &HashMap<H160, T>) -> Vec<TokenPair> {
-    // The `HashMap` docs specifically say that we can't rely on ordering
-    // of keys (even across multiple calls). So, first collect all tokens
-    // into a collection and then use it to make the final enumeration with
-    // all token pair permutations.
-    let tokens = reserves.keys().collect::<Vec<_>>();
-    tokens
-        .iter()
-        .enumerate()
-        .flat_map(|(i, &token_a)| {
-            tokens[i + 1..].iter().map(move |&token_b| {
-                TokenPair::new(*token_a, *token_b).expect("unexpected duplicate key in hash map")
-            })
+pub fn token_pairs<T>(reserves: &BTreeMap<H160, T>) -> Vec<TokenPair> {
+    reserves
+        .keys()
+        .flat_map(|&token_a| {
+            reserves
+                .keys()
+                .filter_map(move |&token_b| match TokenPair::new(token_a, token_b) {
+                    Some(pair) if pair.get().0 == token_a => Some(pair),
+                    _ => None,
+                })
```

Comment:
The pattern matching looks quite confusing.
```suggestion
            reserves.keys().filter_map(move |&token_b| {
                let pair = TokenPair::new(token_a, token_b)?;
                (pair.get().0 == token_a).then_some(pair)
            })

```

Reply:
7d31e635d98917da4d35aeea68dabc42df11e20a